core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -1744,7 +1744,7 @@ public static boolean isValidDecimalValue(@Nullable BigDecimal value, RelDataTyp
     case DECIMAL:
       final int intDigits = value.precision() - value.scale();
       final int maxIntDigits = toType.getPrecision() - toType.getScale();
-      return intDigits <= maxIntDigits;
+      return intDigits <= maxIntDigits && value.scale() == toType.getScale();
```

Comment:
this might be the culpit of the change, I would see this as an obivious change of the original intention

Reply:
@zinking Thanks for your review. 
The method is used to judge can remove the cast operator from literal or not. 
If we just judge the precision. The example:
`cast(15.345 as decimal(5,1)) should be return 15.4 but return 15.345. `
If I have any misunderstanding please let me know.